### ISSUE TITLE
APP-CRASH pytorch (AVX2) illegal instruction under just drrun

### ISSUE THREAD

**Issue by andrewbartolo** [DynamoRIO/dynamorio#4652]
Hi all,

In the spirit of the recent AVX fixes, I've come across what seems to be another issue with AVX2 under just drrun (no client).

My execution environment:
CPU: Xeon E5-2683 v4 (Broadwell, AVX2)
Hypervisor: Proxmox (Debian 10) KVM in cpu-passthrough mode (CPUID shows native CPU in VM)
VM OS: Debian 11 (also tried with Ubuntu 20.04)
DR build: bbf41a8 (latest)
Runtime language: Python 3.9.1
Application: PyTorch 1.7.1

The application works fine without drrun. However, when I invoke `drrun -- python3 resnet-evaluate.py`, drrun stops with `Illegal instruction`. The application also works fine under drrun if I set KVM to emulate a Sandy Bridge (AVX1) CPU instead, as PyTorch auto-detects CPUID features and then runs with the highest-supported level. I'm unsure if PyTorch falls back to AVX1 or all the way back to SSE instructions on Sandy Bridge, but it no longer throws `Illegal instruction`.

I built DR in debug mode, and got the following backtrace:

```
Signal 4 delivered to application as default action.
Callstack:
	0x00007f0c5c661990   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x3a6d990>
	0x00007f0c59916894   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22894>
	0x00007f0c59916d85   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22d85>
	0x00007f0c59909b47   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd15b47>
	0x0000000000000000  
	0x0000000000000001  
	0x00007f0c1b514100   <.../lib/python3.9/site-packages/torch/lib/libc10.so+0x3f100>
	0x002378b8058b4808  
>
<Stopping application /usr/bin/python3.9 (17212)>
Illegal instruction
```

After objdumping libtorch_cpu.so, I found:

```
3a6d985:	0c 0a 
3a6d987:	c5 fc 59 c2          	vmulps %ymm2,%ymm0,%ymm0
3a6d98b:	c4 e2 55 98 c3       	vfmadd132ps %ymm3,%ymm5,%ymm0
3a6d990:	c5 c4 54 2d 68 1d 0c 	vandps 0xa0c1d68(%rip),%ymm7,%ymm5        # db2f700 <_ZL20_ps256_inv_sign_mask>
3a6d997:	0a 
3a6d998:	c4 e2 35 bc c3       	vfnmadd231ps %ymm3,%ymm9,%ymm0
3a6d99d:	c5 fc 58 c2          	vaddps %ymm2,%ymm0,%ymm0
3a6d9a1:	c5 c4 54 3d 77 1d 0c 	vandps 0xa0c1d77(%rip),%ymm7,%ymm7        # db2f720 <_ZL16_ps256_sign_mask>
3a6d9a8:	0a 
```

So, it seems like `vandps` is causing some issue (or possibly `vfmadd132ps` on the previous line?).

I took a look at this line in `core/ir/x86/decode_table.c` together with the `Vvs`, `Hvs`, and `Wvs` definitions:

` {OP_vandps,  0x0f5410, "vandps",  Vvs, xx, Hvs, Wvs, xx, mrm|vex, x, tevexwb[205][0]},`

and all of the source + destination operands and flags seem to match the [Intel reference](https://www.felixcloutier.com/x86/andps). I'm less familiar with the structure of the DR decode table, so maybe something is mixed up elsewhere?

The script I'm running is at [https://pytorch.org/hub/pytorch_vision_resnet/](https://pytorch.org/hub/pytorch_vision_resnet/). I can run the first line (`import torch`), but the next two lines activate the AVX2 instructions and cause the crash.

Happy to assist with any other info -- just wondering if I'm missing anything obvious in the decode table for that instruction.

**Comment by khuey** [comment]
vfmadd132ps is a FMA instruction, not a plain AVX2 instruction. Does your CPU support that?

**Comment by andrewbartolo** [comment]
Yes, it does. Here are the flags from `lscpu` from within the VM:

```
fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse
sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid
tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt
tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault
invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm
rdseed adx smap xsaveopt arat umip md_clear arch_capabilities
```

The application runs fine with all these flags natively, which leads me to believe that the issue is somewhere within DR.

**Comment by khuey** [comment]
Ok. There's nothing obviously wrong with the decoding (running `drdisas` with those byte sequences produces sensible results). Can you figure out what's actually executing when the SIGILL is raised?

**Comment by andrewbartolo** [comment]
I'm not aware of any other threads, etc. that would be executing when SIGILL is raised. I ran the debug build again in gdb, fast-forwarded through the initial frame measurement SIGILL and all the internal SIGBUSes and SIGSEGVs in between, and got the same result:

```
<Application /usr/bin/python3.9 (2435).  Application exception at PC 0x00007ffde80a1990.  
Signal 4 delivered to application as default action.
Callstack:
	0x00007ffde80a1990   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x3a6d990>
	0x00007ffde5356894   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22894>
	0x00007ffde5356d85   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22d85>
	0x00007ffde5349b47   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd15b47>
	0x0000000000000000  
	0x0000000000000001  
	0x00007ffda6f54100   <.../lib/python3.9/site-packages/torch/lib/libc10.so+0x3f100>
	0x002378b8058b4808  
>
<Stopping application /usr/bin/python3.9 (2435)>

Program received signal SIGILL, Illegal instruction.
```

I dug a little more, and it turns out that the same form of that `vandps` instruction occurs several instructions earlier in the same basic block (and seems to execute successfully). Several instances of the three-register (non-memory) `vandps` also occur

```
3a6d8f8:   c5 e4 54 1d 40 1e 0c    vandps 0xa0c1e40(%rip),%ymm3,%ymm3        # db2f740 <_ZL20_ps256_inv_mant_mask>      

...

3a6d91e:   c5 e4 54 d5             vandps %ymm5,%ymm3,%ymm2

...

3a6d990:   c5 c4 54 2d 68 1d 0c    vandps 0xa0c1d68(%rip),%ymm7,%ymm5        # db2f700 <_ZL20_ps256_inv_sign_mask>        
```

which, combined with the fact that `drdisas` worked on the static byte sequences, makes me think that there might be an issue with some as-of-yet-hidden DR state?

**Comment by derekbruening** [comment]
Please run `-debug -loglevel 2` and look for "Got signal at pc 0xxxx in this fragment:" to get the actual instruction that generated the SIGILL.

**Comment by andrewbartolo** [comment]
With `-debug -loglevel 2`, this is from the thread-specific logfile `log.0.17388.html` (there was only one thread running):

```
d_r_dispatch: target = 0x00007ff230dc083d
Fragment 395048, tag 0x00007ff230dc083d, flags 0x9000618, shared, tracehead, size 56:

Entry into F395048(0x00007ff230dc083d).0x000000004b23d7d4 (trace head)(shared)
Exit from F395048(0x00007ff230dc083d).0x000000004b23d807 (shared)
 (target 0x00007ff230dc0870 not in cache)

d_r_dispatch: target = 0x00007ff230dc0870
interp: invalid instr at 0x00007ff230dc0990
Fragment 395049, tag 0x00007ff230dc0870, flags 0x1000630, shared, size 881:

Entry into F395049(0x00007ff230dc0870).0x000000004b28fde0 (shared)
Exit from F395049(0x00007ff230dc0870).0x000000004b29014c (shared)
 (target 0x00007ff230dc0990 not in cache)

d_r_dispatch: target = 0x00007ff230dc0990
interp: invalid instr at 0x00007ff230dc0990
Variable-size block: allocating 0x00007ff240e12050 (896 bytes [896 aligned] in 131080 block)
record_pending_signal(4) from DR at pc 0x00007ff230dc0990
transfer_to_dispatch: pc=0x30dc0990, xsp=0x00007ffc3729efa0, on-initstack=0
Exit from asynch event
execute_handler_from_dispatch for signal 4
app signal handler is SIG_DFL: executing default action
SYSLOG_WARNING: Application /usr/bin/python3.9 (17388).  Application exception at PC 0x00007ff230dc0990.
Signal 4 delivered to application as default action.
Callstack:
    0x00007ff230dc0990   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x3a6d990>
    0x00007ff22e075894   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22894>
    0x00007ff22e075d85   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd22d85>
    0x00007ff22e068b47   <.../lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xd15b47>
    0x0000000000000000
    0x0000000000000001
    0x00007ff1efc73100   <.../lib/python3.9/site-packages/torch/lib/libc10.so+0x3f100>
    0x002378b8058b4808

Terminating via kill
SYSLOG_INFORMATION: Stopping application /usr/bin/python3.9 (17388)
```

Is this not indicating that the relevant instruction is at `libtorch_cpu.so+0x3a6d990`? Should I check `/proc/$PID/maps` instead?

Let me know, happy to help.

**Comment by derekbruening** [comment]
Ah, `-loglevel 2` would do it for a SIGSEGV or sthg that needs translation -- but not for a SIGILL discovered in bb building.  `-loglevel 3` will print the raw bytes that DR's decoder thinks is invalid.  Though that offset 0x990 does seem to point at what you were already looking at based on the printed callstack: does the log at level 3 show the exact same byte sequence `c5 c4 54 2d 68 1d 0c 0a`?

**Comment by andrewbartolo** [comment]
Yep, looks like the same sequence.

```
 0x00007f8152d82987  c5 fc 59 c2          vmulps %ymm0 %ymm2 -> %ymm0
 0x00007f8152d8298b  c4 e2 55 98 c3       vfmadd132ps %ymm5 %ymm3 %ymm0 -> %ymm0
 0x00007f8152d82990  c5 c4 54 2d 68 1d 0c vandps %ymm7 <rel> 0x00007f815ce44700[32byte] -> %ymm5
                     0a
interp:invalid instr at 0x00007f8152d82990
end_pc= 0x00007f8152d82990

exit_branch_type=0x0bb->exit_target=0x00007f8152d82990
exit_branch_type=0x0target=0x00007f8152d82990 l->flags=0x9801
Fragment395040, tag 0x00007f8152d82870, flags 0x1000630, shared, size 881:

Entryinto F395040(0x00007f8152d82870).0x00000000486d57e0 (shared)

Exitfrom F395040(0x00007f8152d82870).0x00000000486d5b4c (shared)
(target 0x00007f8152d82990 not in cache)

d_r_dispatch:target = 0x00007f8152d82990

interp:start_pc = 0x00007f8152d82990
 0x00007f8152d82990  c5 c4 54 2d 68 1d 0c vandps %ymm7 <rel> 0x00007f815ce44700[32byte] -> %ymm5
                     0a
interp:invalid instr at 0x00007f8152d82990
Variable-sizeblock: allocating 0x00007f8162dbe050 (896 bytes [896 aligned] in 131080 block)
record_pending_signal(4)from DR at pc 0x00007f8152d82990
   action is not SIG_IGN
   retaddr = 0x00007f8362b03e4c
copy_frame_to_pendingfrom 0x00007f81628dd880
```

**Comment by derekbruening** [comment]
Looks like decode_cti() fails to decode it, yet the full decoder decodes it just fine, which is why drdisas has no trouble:

```
$ clients/bin64/drdisas c5 c4 54 2d 68 1d 0c 0a 90
 c5 c4 54 2d 68 1d 0c vandps ymm5, ymm7, <rel> yword ptr [0x000055b3271b4e60]
 0a
 90                   nop
```

Issues:
+ What exactly is the bug, of course.
+ Presumably decode_sizeof succeeds since it knows the length?
+ We have decode_cti tests for one or a few operand or size variants of every opcodes: but not of any from-bits patterns.  Should we add decode_cti to the binutils tests: would that help?

The workaround would be to pass `-full_decode`.  @andrewbartolo : can you confirm that solves the problem?

**Comment by andrewbartolo** [comment]
Running with `drrun -debug -loglevel 3 -full_decode -- python3 resnet-evaluate.py` led to a different, DR-internal SIGSEGV crash. The stack trace wasn't printed, but I looked at the `-loglevel 3` logfile and the final disassembled fragment printed in the logfile is the same one that contains the `vandps`.

Here is the log output indicating the crash and (cache) pc:

```
computing memory target for 0x00007fbc86a6fd22 causing SIGSEGV, kernel claims it is 0x0000000000000000
26285182 For SIGSEGV at cache pc 0x00007fbc86a6fd22, computed target read 0x0000000000000000
26285183     faulting instr: movzx  (%rax)[1byte] -> %eax
26285184 SYSLOG_CRITICAL: Application /usr/bin/python3.9 (19460).  DynamoRIO internal crash at PC 0x00007fbc86a6fd22.  Please report this at http://dynamorio.org/issues/.  Program aborted.
26285185 Received SIGSEGV at pc 0x00007fbc86a6fd22 in thread 19460
```

**Comment by derekbruening** [comment]
The full fragment would be needed to tell where that movzx came from.  Ideally the whole sequence of the original block and the instrlist-level disasm post-mangling which labels inserted vs original instrs.  It may well be unrelated to this issue and opened in a separate issue once diagnosed.

**Comment by derekbruening** [comment]
@khuey wondering if you are able to look at the decode_cti problem?

**Comment by khuey** [comment]
Did we establish exactly what byte sequence is causing decode_cti to fail?

**Comment by derekbruening** [comment]
Yes, the original sequence, echoed in https://github.com/DynamoRIO/dynamorio/issues/4652#issuecomment-753730645: `c5 c4 54 2d 68 1d 0c 0a`

**Comment by khuey** [comment]
Ok, I'll take a look when I have some time later this week.

**Comment by khuey** [comment]
It seems possible but far from guaranteed that #4667 will fix the original issue here.

**Comment by andrewbartolo** [comment]
I just tested and can confirm that #4667 fixes both the `decode_cti()` issue and also the DR-internal crash I saw for `-debug -full_decode`. (So, I guess `decode_cti()` is part of `-full_decode` somehow as well.)

Thanks @khuey and @derekbruening for your work on this!
